Configurable spatial index

craigtaverner opened the pull request:
* Moved spatial index settings to separate file
* Make extents configurable
* Saving and reading index settings

craigtaverner committed:
Configurable spatial index

* Moved spatial index settings to separate file
* Make extents configurable
* Saving and reading index settings

sherfert committed:
Move half-open range computation to space filling curve.

craigtaverner committed:
Descriptions for spatial index extents settings

OliviaYtterbrink commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialCRSSchemaIndex.java:
```diff
@@ -327,26 +331,62 @@ public boolean indexExists()
         return NativeSchemaIndexes.readState( pageCache, indexFile, layout );
     }
 
+    private void checkLayout() throws IOException
```
Very surprising to get side effects from this method...

OliviaYtterbrink commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/config/SpaceFillingCurveSettings.java:
```diff
@@ -0,0 +1,170 @@
+    /**
+     * @return The number of dimensions (2D or 3D)
+     */
+    public int getDimensions()
```
The following three getters are never used.

OliviaYtterbrink commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/config/SpaceFillingCurveSettingsFactory.java:
```diff
@@ -0,0 +1,155 @@
+                String[] fields = key.replace( SPATIAL_SETTING_PREFIX, "" ).split( "\\." );
+                CoordinateReferenceSystem crs = CoordinateReferenceSystem.byName( fields[0] );
+                EnvelopeSettings envelopeSettings = env.computeIfAbsent( crs, EnvelopeSettings::new );
+                int index = "xyz".indexOf( fields[1].toLowerCase() );
```
Maybe should check the size of fields before accessing here, could result in out of bounds, at least when adding support for custom crs.

craigtaverner committed:
Port spatial index settings into GBPTree header instead of separate file

OliviaYtterbrink reviewed (commented):
Mostly nice, some pain points like increased complexity of `SpatialCRSSSchemaIndex` (though that will get removed in the refactoring), maybe some better errorhandling for badly specified settings and tests for `SpaceFillingCurveSettings(Factory)`

craigtaverner committed:
Fixed Spatial index config settings support after fixes to header writer

craigtaverner committed:
Got configurable spatial index working on new index framework

OliviaYtterbrink committed:
Don't close the populator multiple times

Populator was always marked as failed when running `withPopulator`
which overwrites the header, leading to the spatial index being
unable to read settings in the GBPtree.
